atlasdb-timelock-server/src/main/java/com/palantir/atlasdb/timelock/paxos/PaxosTimestampBoundStore.java

```diff
@@ -0,0 +1,286 @@
+        if (!value.getLeaderUUID().equals(proposer.getUUID())) {
+            String errorMsg = String.format(
+                    "Timestamp limit changed from under us for sequence '%s' (leader with UUID '%s' changed"
+                            + " it, our UUID is '%s'). This suggests that another timestamp store for this"
+                            + " namespace is running. The offending bound was '%s'; we tried to propose"
+                            + " a bound of '%s'. (The offending Paxos value was '%s'.)",
+                    newSeq,
+                    value.getLeaderUUID(),
+                    proposer.getUUID(),
+                    PtBytes.toLong(value.getData()),
+                    limit,
+                    value);
+            throw new MultipleRunningTimestampServiceError(errorMsg);
+        }
+    }
+
+    /**
+     * Executes a backoff action which is given a random amount of time to wait in milliseconds. This is used in Paxos
+     * to resolve multiple concurrent proposals. Users are allowed to specify their own backoff action,
+     * to handle cases where users hold or do not hold monitor locks, for instance.
+     *
+     * @param paxosException the PaxosRoundFailureException that caused us to wait
+     * @param backoffAction the action to take (which consumes the time, in milliseconds, to wait for)
+     */
+    private void waitForRandomBackoff(PaxosRoundFailureException paxosException, BackoffAction backoffAction) {
+        long backoffTime = getRandomBackoffTime();
+        log.info("Paxos proposal couldn't complete, because we could not connect to a quorum of nodes. We"
+                + " will retry in {}ms. The Paxos exception was {}",
+                backoffTime,
+                paxosException);
+        try {
+            backoffAction.accept(backoffTime);
+        } catch (InterruptedException interruptedException) {
+            Thread.currentThread().interrupt();
+        }
+    }
+
+    /**
+     * Generates a random amount of time to wait for, in milliseconds.
+     * This typically depends on the configuration of the Paxos algorithm; currently, we have implemented
+     * this as U(0, k) where k is the maximum wait before proposal in the Paxos configuration.
+     *
+     * @return the amount of time to wait for, in milliseconds
+     */
+    private long getRandomBackoffTime() {
+        return (long) (maximumWaitBeforeProposalMs * Math.random());
+    }
+
+    @Value.Immutable
+    interface PaxosLong extends PaxosResponse {
+        @Override
+        default boolean isSuccessful() {
+            return true;
+        }
+
+        @Value.Parameter
+        long getValue();
+    }
+
+    @Value.Immutable
+    interface SequenceAndBound {
+        @Value.Parameter
+        long getSeqId();
+
+        @Value.Parameter
+        long getBound();
+    }
+
+    private interface BackoffAction {
+        void accept(long backoffTime) throws InterruptedException;
```

Comment:
Okay. We do need the interface because the retry methods throw a checked exception; there isn't direct support for this, it seems, and [solutions](http://stackoverflow.com/questions/27644361/how-can-i-throw-checked-exceptions-from-inside-java-8-streams) generally involve defining one's own functional interfaces.

I used `accept` for consistency with Java's `Consumer` which is what this class is essentially supposed to be (if not for the `InterruptedException` I would have just used a `Consumer<Long>`. That said I agree that overloading the paxos terminology can quickly get confusing, so I've renamed the method to `backoff`.